File: Leetcode/Javascript/200. Number of Islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    var count = 0;
    for(var i = 0; i < grid.length; ++i) {
        for(var j = 0; j < grid[0].length; ++j) {

            if(grid[i][j] <= 0) continue;
            dfs(i,j,grid);
            count++;
        }
    }
    return count;
};

function dfs(i, j, grid) {
    if(i < 0 || j < 0 || i == grid.length || j == grid.length || grid[i][j] <= 0) return;
    grid[i][j] = -1;
    dfs(i-1, j, grid);
    dfs(i+1, j,grid);
    dfs(i, j-1, grid);
    dfs(i, j+1, grid);
}
/*
    Js的test case有问题...似乎不是按照2D-grid的方式来写的
*/

File: Leetcode/One/200. Number of Islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int height = grid.length;
        if(height == 0) return 0;
        int res = 0;
        int width = grid[0].length;
        for(int i = 0; i < height; ++i) {
            for(int j = 0; j < width; ++j) {
                if(!isLand(grid[i][j])) continue;
                dfs(grid, i, j);
                res++;
            }
        }
        return res;
    }
    private void dfs(char[][] grid, int x, int y) {
        int height = grid.length,
            width = grid[0].length;
        if(x < 0 || x == height || y < 0 || y == width || !isLand(grid[x][y])) return;
        grid[x][y] = '0';
        dfs(grid, x-1, y);
        dfs(grid, x+1, y);
        dfs(grid, x, y-1);
        dfs(grid, x, y+1);
    }
    private boolean isLand(char cell) {
      return cell == '1';
    }
}

class Solution {
    public int numIslands(char[][] grid) {
        int height = grid.length;
        if(height == 0) return 0;
        int res = 0;
        int width = grid[0].length;
        for(int i = 0; i < height; ++i) {
            for(int j = 0; j < width; ++j) {
                if(!isLand(grid[i][j])) continue;
                bfs(grid, i, j);
                res++;
            }
        }
        return res;
    }

    private void bfs(char[][] grid, int x, int y) {
        Queue<Character> queue = new LinkedList<>();
        int height = grid.length,
            width = grid[0].length;
        for(int i = 0; i < height; ++i) {
            for(int j = 0; j < width; ++j) {

            }
        }
    }

    private boolean isLand(char cell) {
      return cell == '1';
    }
}
